hazelcast/src/test/java/com/hazelcast/map/impl/tx/TxnMapNearCacheTest.java

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.tx;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.config.NearCacheConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.test.HazelcastParametersRunnerFactory;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.transaction.TransactionException;
+import com.hazelcast.transaction.TransactionalTask;
+import com.hazelcast.transaction.TransactionalTaskContext;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.map.impl.tx.TxnMapNearCacheTest.InvalidatorTxnOp.DELETE;
+import static com.hazelcast.map.impl.tx.TxnMapNearCacheTest.InvalidatorTxnOp.PUT;
+import static com.hazelcast.map.impl.tx.TxnMapNearCacheTest.InvalidatorTxnOp.PUT_IF_ABSENT;
+import static com.hazelcast.map.impl.tx.TxnMapNearCacheTest.InvalidatorTxnOp.PUT_TTL;
+import static com.hazelcast.map.impl.tx.TxnMapNearCacheTest.InvalidatorTxnOp.REPLACE;
+import static com.hazelcast.map.impl.tx.TxnMapNearCacheTest.InvalidatorTxnOp.REPLACE_IF_SAME;
+import static com.hazelcast.map.impl.tx.TxnMapNearCacheTest.InvalidatorTxnOp.SET;
+import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(HazelcastParametersRunnerFactory.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class TxnMapNearCacheTest extends HazelcastTestSupport {
```

Comment:
I somehow find this test hard to read. Maybe it's due to the jumping between the test and `enum` and between the `enum `overrides and the default methods (actually I didn't even know that an `enum` can have abstract methods).

Maybe an old school `switch` is easier here. You could re-use the enum [`DataStructureAdapter.DataStructureMethods`](https://github.com/hazelcast/hazelcast/blob/7aa4eb19cf3142cdd7daf2ea16491ee63656ec14/hazelcast/src/main/java/com/hazelcast/internal/adapter/DataStructureAdapter.java#L125). This will also get rid of the code duplicate for `DELETE`, `REPLACE` and `REPLACE_IF_SAME` (assuming that it doesn't matter if `set()` or `put()` is used), e.g.:

```java
private void txnInvalidatesNearCache(final DataStructureMethods method) throws Exception {
    ...

    // 1. populate the Near Cache
    switch (method) {
        case DELETE:
        case REPLACE:
        case REPLACE_WITH_OLD_VALUE:
            for (int i = 0; i < numberOfEntries; i++) {
                map.set(i, i);
            }
            for (int i = 0; i < numberOfEntries; i++) {
                map.get(i);
            }
            break;
        default:
    	    // populate Near Cache with nulls
            for (int i = 0; i < numberOfEntries; i++) {
                map.get(i);
            }
    }

    // 2. run a txn task which causes invalidation
    TransactionalTask<Object> task = new TransactionalTask<Object>() {
        @Override
        public Object execute(TransactionalTaskContext context) throws TransactionException {
            for (int i = 0; i < numberOfEntries; i++) {
                switch (method) {
                    case SET:
                        context.getMap(mapName).set(i, i);
                        break;
                    case ...
                        break;
                }
            }
            return null;
        }
    };
    node1.executeTransaction(task);

    // 3. assert
    switch (method) {
         ...
    }
}
```
The three blocks could be pulled out as methods. In the end this will still be linear to read (just down, no jumping up should be needed).